Create route for single version

lightandluck opened the pull request:
Fixes #145 

Added logic to handle `/page/page_uuid/..versionId` so that it will show the change between the provided version and the one previous to it.
Added the new route, which is a component that builds a redirect to that path.
Also had to add a function to the db api to look up a version by `versionId` so that I could get the `page_uuid` to build the route.

Mr0grog commented on src/components/version-redirect.jsx:
```diff
@@ -0,0 +1,25 @@
+    const versionId = this.props.match.params.versionId;
+    this.props.api.getVersionById(versionId).then(data => {
+      this.setState({pageId: data.page_uuid});
+    });
```
It might be simpler to just call `this.props.history.replace()` instead of setting state here, e.g:

```js
this.props.api.getVersionById(versionId).then(data =>
  this.props.history.replace(`/page/${data.page_uuid}/..${versionId}`));
```

Then you don’t need the constructor at all and the `render()` method only needs to always return `null` because there’s no state to track.

(Side note: it’s pretty edge-case-y, but you could check that the `versionId` in the props still matches the one you initially had before redirecting… in case the request was really slow and the user did something to navigate to a new version in the mean time.)

lightandluck replied:
Cool, will change. I didn't know I could do that with `history`. I thought I always needed a `<Redirect />`. If all you have is a hammer...

Mr0grog replied:
Yeah. If you didn’t have to make an async API call, the `<Redirect>` would probably be clearer and more declarative.

Now that I’m thinking more about it, though… maybe you do need to have some state here so that you can render:

- The loading indicator while checking for the version data,
- An error/not found message if the version ID is bogus

…at that point, maybe the `<Redirect>` is just fine ¯\\\_(ツ)\_/¯

lightandluck commented on src/components/page-details.jsx:
```diff
@@ -149,9 +149,9 @@ export default class PageDetails extends React.Component {
     if (!versionData) {
-      let [to, from] = this.state.page.versions;
-      from = from || to;
+      let [to, from] = this._parsedVersionsToRender() || this.state.page.versions;
```
Yea, I was really struggling here when writing. I initially did change `_versionsToRender` but wasn't sure what to return to make use of the logic in `_renderChange` without a bunch of repeated code and crazy conditionals. So, I settled on just creating a new function and plugging it into that line and not have to change any surrounding code.

I do like this a lot more without having the additional function. And using `find` like that to get the previous relative version is real nifty! Will change.

lightandluck commented:
@Mr0grog Good point about loading and error handling. I added those and implemented the recommended changes in `page-details`. 

Sorry to muddle things a bit, but I also fixed up `<Loading />` to use flexbox for consistency while I was at it. Looking at it more carefully now, I just couldn't leave it be. Sorry!

Let me know how things look.

Mr0grog commented on src/components/web-monitoring-ui.jsx:
```diff
@@ -166,6 +167,7 @@ export default class WebMonitoringUi extends React.Component {
                 pages={this.state[this.state.pageFilter]}
               />
             }/>
+            <Route path="/version/:versionId" render={(routeProps) => <VersionRedirect api={api} {...routeProps} />} />
```
Somehow I totally missed this before. `VersionRedirect` should be getting the API object via its `context`, not an explicit prop. Same as all the other routes.

Since you don’t need to add anything to it, I think you can use `component` instead of `render`, too:

```jsx
<Route path="/version/:versionId" component={VersionRedirect} />
```

Mr0grog commented on src/components/version-redirect.jsx:
```diff
@@ -0,0 +1,36 @@
+
+  componentDidMount () {
+    const versionId = this.props.match.params.versionId;
+    this.props.api.getVersionById(versionId)
```
I completely missed this before, sorry. `api` should be coming from `context`, where it’s already available, not passed as an explicit prop. This should be the same as all the other components that use it.

Mr0grog commented on src/services/web-monitoring-db.js:
```diff
@@ -207,6 +207,17 @@ export default class WebMonitoringDb {
+      .then(data => parseVersion(data.data));
+  }
+
+  /**
```
I don’t think you should add this method. You should remove the `pageId` parameter from `getVersion` (it predates the ability to actually get a version without a page ID).

(Sorry again for missing so many things before. I should not have done a midnight review.)

Mr0grog commented on src/components/version-redirect.jsx:
```diff
@@ -0,0 +1,36 @@
+      return (
+        <p className="alert alert-danger" role="alert">
+          Error: We couldn't find the version you're looking for.
+          Please check you provided the correct versionID or alert the dev team.
```
I don’t think “or alert the dev team” makes sense in the larger context if someone else redeploys this or as we add more external users in the future. Any EDGI analysts know they can just bug us on Slack.

lightandluck commented:
No problem. Sorry, I forgot all about `context`. Made the switch and was able to simplify the route to just use `component`. 

I also added this [line](https://github.com/lightandluck/web-monitoring-ui/blob/14a35e592cdfdc79a5029f1b6ad7b1e95ff13e5c/src/components/version-redirect.jsx#L20) to check for the same version ID before redirecting. I believe that a closure is created and when the promise resolves, `versionId` will hold the original ID and `props` will hold the current one if anything had changed. It gets rid of `trying to setState() on unmounted component` errors I was getting when navigating away from the page before the promise resolves. But I may be misunderstanding why that gets rid of the error.

I searched around and there were suggestions to create an `isMounted` variable in state, setting that to `false` in `componentWillUnmount`,  and checking that. I could go that direction if that handles the situation better. I'm uncertain.

Noted about the extra work on Loading. Will open a new PR next time!

Mr0grog commented:
> I also added this line to check for the same version ID before redirecting… It gets rid of trying to setState() on unmounted component errors I was getting when navigating away from the page before the promise resolves. But I may be misunderstanding why that gets rid of the error.

You’re understanding exactly right. This is what I was talking when I said:

> (Side note: it’s pretty edge-case-y, but you could check that the versionId in the props still matches the one you initially had before redirecting… in case the request was really slow and the user did something to navigate to a new version in the mean time.)

> there were suggestions to create an isMounted variable in state, setting that to false in componentWillUnmount, and checking that. I could go that direction if that handles the situation better. I'm uncertain.

*Technically speaking,* your fix and that one are addressing two different problems (a race condition with a change in props vs. a race condition with a change in mounting). However, now that you’ve gone back to routing via rendering (instead of directly touching the history API), you shouldn’t have to worry about the mounting race condition—updating state should do nothing if the component is unmounted.

Mr0grog commented:
Anyway, looks good to me! 👍

lightandluck commented:
Cool, yea, I was trying to address that edge case because I had forgotten to. Glad it worked. Thanks for the explanation, it makes a lot more sense now. 

In it goes!
